File: 0000-0999/0210.CourseScheduleII.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        graph = {i:[] for i in xrange(numCourses)}
        inDegree = {i:0 for i in xrange(numCourses)}
        for ch, fa in prerequisites:
            graph[fa].append(ch)
            inDegree[ch] += 1
        topo = []
        for course in xrange(numCourses):
            if inDegree[course] == 0:
                topo.append(course)
        i = 0
        while i < len(topo):
            for course in graph[topo[i]]:
                inDegree[course] -= 1
                if inDegree[course] == 0:
                    topo.append(course)
            i += 1
        if len(topo) == numCourses:
            return topo
        return []

File: 0000-0999/0210.CourseScheduleII.go
func findOrder(numCourses int, prerequisites [][]int) []int {
    graph := make([][]int, numCourses)
    indegree := make([]int, numCourses)
    for _, p := range prerequisites {
        graph[p[1]] = append(graph[p[1]], p[0])
        indegree[p[0]]++
    }
    course := make([]int, 0, numCourses)
    for i, d := range indegree {
        if d == 0 {
            course = append(course, i)
        }
    }
    for i := 0; i < len(course); i++ {
        for _, j := range graph[course[i]] {
            indegree[j]--
            if indegree[j] == 0 {
                course = append(course, j)
            }
        }
    }
    if len(course) == numCourses {
        return course
    }
    return nil
}